File: EPI/Searching/Compute the real square root/swift/12.6.swift
public func squareRoot(_ value: Float, tolerance: Float) -> Float {
    let toleranceSquared = tolerance * tolerance
    var (low, high) = value < 1 ? (value, 1) : (1, value)

    while true {
        let mid = low + (high - low) / 2
        let midSquared = mid * mid

        if abs(midSquared - value) < toleranceSquared {
            return mid
        }

        if midSquared < value {
            low = mid
        } else {
            high = mid
        }
    }
}

File: EPI/Searching/Compute the real square root/typescript/12.6.ts
export default function sqrt(x: number, tolerance = 0.01): number {
    const toleranceSquared = tolerance * tolerance;
    let [min, max] = x < 1 ? [x, 1] : [1, x];

    while (true) {
        const mid = min + (max - min) / 2;
        const midSquared = mid * mid;

        if (Math.abs(midSquared - x) < toleranceSquared) {
            return mid;
        }

        if (midSquared < x) {
            min = mid;
        } else {
            max = mid;
        }
    }
}
